libbpf-tools/schedsnoop.c

```diff
@@ -0,0 +1,309 @@
+#include <sys/resource.h>
+#include <sys/wait.h>
+#include <dirent.h>
+#include <bpf/bpf.h>
+#include "errno_helpers.h"
+#include "trace_helpers.h"
+#include "syscall_helpers.h"
+#include "schedsnoop.h"
+#include "schedsnoop.skel.h"
+
+#define NS_IN_SEC		1000000000LLU
+#define NS_IN_MS		1000000LLU
+#define NS_IN_US		1000LLU
+#define PERF_BUFFER_PAGES	64
+
+const char *argp_program_version = "schedsnoop 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+static const char argp_program_doc[] =
+"Trace the related schedule events of a specified task.\n"
+"\n"
+"USAGE: schedsnoop -p PID [-s]\n"
+"\n"
+"EXAMPLES:\n"
+"    schedsnoop -p 49870       	# trace pid 49870\n"
+"    schedsnoop -p 49870 -s    	# trace pid 49870 and system call\n";
+
+static const struct argp_option opts[] = {
+	{ "pid", 'p', "PID", 0, "Process PID to trace"},
+	{ "syscall", 's', NULL, 0, "Trace SYSCALL Info"},
+	{},
+};
+
+static struct env {
+	int nr_cpus;
+	int trace_syscall;
+	int target;
+} env;
+
+int ti_map_fd;
+int si_map_fd;
+int start_map_fd;
+int end_map_fd;
+bool exiting = false;
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	int pid;
+	switch (key) {
+	case 'p':
+		if(arg == NULL) {
+			fprintf(stderr, "PID is required\n");
+			argp_usage(state);
+		} else {
+			pid = strtol(arg, NULL, 10);
+			if (pid <= 0) {
+				fprintf(stderr, "Invalid PID: %s\n", arg);
+				argp_usage(state);
+			}
+			env.target = pid;
+		}
+		break;
+	case 's':
+		env.trace_syscall = 1;
+		break;
+	case ARGP_KEY_END:
+		if (!env.target) {
+			fprintf(stderr, "No target PID\n");
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static inline int time_to_str(u64 ns, char *buf, size_t len)
+{
+
+	if (ns > 10 * NS_IN_SEC)
+		snprintf(buf, len, "%llus", ns / NS_IN_SEC);
+	else if (ns > 10 * NS_IN_MS)
+		snprintf(buf, len, "%llums", ns / NS_IN_MS);
+	else if (ns > 10 * NS_IN_US)
+		snprintf(buf, len, "%lluus", ns / NS_IN_US);
+	else
+		snprintf(buf, len, "%lluns", ns);
+
+	return 0;
+}
+
+static inline void pr_ti(struct trace_info *ti, char *opt, char *delay)
+{
+	char buf[27];
+	snprintf(buf, sizeof(buf), "%lluus", ti->ts / NS_IN_US);
+	printf("%-27sCPU=%-7dPID=%-7dCOMM=%-20s%-37s%-17s\n",
+				buf, ti->cpu, ti->pid, ti->comm, opt,
+				delay ? delay : "");
+}
+
+void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	struct trace_info *ti = data;
+	char d_str[16];
+	char comm_buf[2 * TASK_COMM_LEN];
+	char func[80];
+	struct si_key sik;
+	u64 siv;
+	static u64 w_start, p_start, last_time;
+
+
+	time_to_str(ti->ts - last_time, d_str, sizeof(d_str));
+
+	switch (ti->type) {
+	case TYPE_MIGRATE:
+		w_start = p_start = ti->ts;
+		pr_ti(ti, "MIGRATE", NULL);
+		break;
+	case TYPE_ENQUEUE:
+		w_start = p_start = ti->ts;
+		printf("----------------------------\n");
+		pr_ti(ti, "ENQUEUE", NULL);
+		break;
+	case TYPE_WAIT:
+		if (ti->pid == env.target) {
+			w_start = ti->ts;

```

Comment:
you will actually have a problem with re-ordering of events. If they happen on two different CPUs (e.g., when task is migrated) you can get events out of order, because perfbuf has per-CPU buffers.

I ignored that problem initially, because I assumed all the events are independent, but now I see that you do care about those timestamps.

Luckily, I actually have an even better way to do this for you, except it is quite bleeding edge (you'll need a 5.8 kernel). Do you mind trying that out?

See: https://github.com/torvalds/linux/blob/master/tools/testing/selftests/bpf/progs/test_ringbuf.c and https://github.com/torvalds/linux/blob/master/tools/testing/selftests/bpf/prog_tests/ringbuf.c

BPF_MAP_TYPE_RINGBUF is a shared ringbuffer across all CPUs, so ordering like scheduling events are preserved. It also has a nice reserve/commit API, which seems rather convenient for you here.

Reply:
Well, we do care about the order. we will run some experiments so see how serious the out of order problem is. 

If the perfbuf has per-CPU buffers, does it means that there are per-CPU handlers? E.G. if both cpu#0 and cpu#1 emit an event to the their buffers, will these two events be processed 
simultaneously by related cpu?

And about the shared ringbuffer in kernel 5.8, we are very interesting in this new feature. However, we are worried whether this feature is stable or not. Will the mainline kernel 5.8 include this feature in the future? If it does, we are very willing to have a try

Reply:
I used to trace fork/exec/exit events, and re-ordering happened regularly enough to be a big problem. For you it will mostly be happening when thread is migrated to a different CPU and keeps running there.

re: per-CPU buffers. Each buffer is independent, but your user-space callback is going to be executed on single CPU. So reordering happens only because you might get events from CPU X before events from CPU Y, even though some of the events on CPU Y happened sooner.

re: BPF RINGBUF, yes, it's stable. If there are any bugs, we'll be fixing them, but so far we haven't seen any report. It's extremely unlikely this will be pulled out. We are going to use it at FB in multiple applications, so I think you can definitely use it.